Add EXPLAIN (FORMAT tscout) (extension "Hutch") to display features for the behavior models.

karthik-ramanathan-3006 opened the pull request:
Add EXPLAIN (FORMAT tscout) (extension "Hutch") to display features for the behavior models.

This PR modifies core PostgreSQL to add the `tscout` EXPLAIN format.
By default, `EXPLAIN (FORMAT tscout)` will behave exactly like `EXPLAIN (FORMAT json)`.
However, when the extension is loaded, the output is prepended with augmented JSON
that contains all of the features used by the operating units (see sample output below).

To install this extension,
```
cd cmudb/tscout/extensions/tscout/
sudo python3 tscout_feature_gen.py
make clean -j
make -j
make install -j
# This defaults to the default `build/` directory specified by `cmudb/build/configure.sh`.
# Specify a different PostgreSQL install by passing the PG_CONFIG env var
# to make.
```
For convenience, a CMake target `extensions_tscout_install` will do these steps
and install this extension into the default `build/` directory.

To load this extension, either method of loading extensions is fine:
- Run `LOAD 'hutch_extension'`.
    - This loads the extension for the current connection.
- Add `shared_preload_libraries = 'hutch_extension'` to `postgresql.conf`.
    - This loads the extension for all future connections.

Sample `psql` output:
```
foo=# LOAD 'tscout_extension';
2022-01-30 00:39:22.367 EST [35725] LOG:  Initializing extension.
2022-01-30 00:39:22.367 EST [35725] STATEMENT:  LOAD 'tscout_extension';
LOAD
foo=# EXPLAIN (FORMAT TSCOUT) SELECT 1;
                 QUERY PLAN                 
--------------------------------------------
 [                                         +
   {                                       +
     "node": "node-0",                     +
     "tag": "ExecResult",                  +
     "Result_plan_type": 10,               +
     "Result_plan_startup_cost": 0.000000, +
     "Result_plan_total_cost": 0.010000,   +
     "Result_plan_plan_rows": 1.000000,    +
     "Result_plan_plan_width": 4,          +
     "Result_plan_parallel_aware": false,  +
     "Result_plan_parallel_safe": false,   +
     "Result_plan_async_capable": false,   +
     "Result_plan_plan_node_id": 0,        +
     "Result_plan_targetlist": "<skipped>",+
     "Result_plan_qual": "<skipped>",      +
     "Result_plan_lefttree": "<skipped>",  +
     "Result_plan_righttree": "<skipped>", +
     "Result_plan_initPlan": "<skipped>",  +
     "Result_plan_extParam": "<skipped>",  +
     "Result_plan_allParam": "<skipped>",  +
     "Result_resconstantqual": "<skipped>" +
   },                                      +
   {                                       +
     "Plan": {                             +
       "Node Type": "Result",              +
       "Parallel Aware": false,            +
       "Async Capable": false,             +
       "Startup Cost": 0.00,               +
       "Total Cost": 0.01,                 +
       "Plan Rows": 1,                     +
       "Plan Width": 4                     +
     }                                     +
   }                                       +
 ]
(1 row)
```

https://www.youtube.com/watch?v=r8ZC9uDg8Ik

---

Adds a new option to EXPLAIN command: "tscout"

Run the following commands to "make" the extension:
cd cmudb/extensions/hutch/
make clean; make; make install

karthik-ramanathan-3006 committed:
Adds skeleton for a Postgres TScout extension.
Adds a new option to EXPLAIN command: "tscout"

Run the following commands to "make" the extension:
cd cmudb/extensions/tscout/
make clean; make; make install

The extension is imported as a shared library (.so file) during server startup.
TODO: Automate this.

karthik-ramanathan-3006 committed:
Adds functionality to the extension to generate Xs and plan related values.

karthik-ramanathan-3006 committed:
Adds functionality to the extension to scrape values related to the Xs, autogenerates supporting infrastructure.

karthik-ramanathan-3006 committed:
Merge upstream branch 'pg14' into tscout_ext

karthik-ramanathan-3006 committed:
Run clang-format.

Signed-off-by: Wan Shen Lim <wanshen.lim@gmail.com>

lmwnshn committed:
Fix include issues caused by clang-format.

Signed-off-by: Wan Shen Lim <wanshen.lim@gmail.com>

lmwnshn committed:
Merge branch 'pg14' of github.com:cmu-db/postgres into karthik_tscout_ext

lmwnshn committed:
Set a default location for PG_CONFIG.

lmwnshn committed:
Run `isort` on Python files.

lmwnshn committed:
Add extensions/tscout/*.[ch] to `format` target.

lmwnshn committed:
Run `format`.

lmwnshn committed:
Modify POSTGRES_PATH of clang_parser to remove dependency on working directory.

lmwnshn committed:
Increase verbosity of .gitignore comment.

lmwnshn committed:
Add extensions_tscout_install CMake target.

See `cmudb/extensions/tscout/CMakeLists.txt` for details.

lmwnshn committed:
Fix typo.

lmwnshn committed:
Nitpick spelling.

lmwnshn commented on cmudb/tscout/model.py:
```diff
@@ -677,3 +677,13 @@ def __init__(self):
+        self._enums = nodes.enum_map
+
+    def get_enum_value_map(self, enum_name):
+        if enum_name not in self._enums.keys():
```
It appears that this code path is never hit; I replaced it with a `raise Exception` and nothing seemed to break, at least.

Given that, please consider adding an assertion instead. The idea is that it is better to fail directly and loudly as soon as possible, rather than work with questionable data.

karthik-ramanathan-3006 replied:
Good point! Makes sense. Thanks!

lmwnshn commented on cmudb/extensions/tscout/main.c:
```diff
@@ -0,0 +1,224 @@
+#include "parser/parsetree.h"
+#include "utils/builtins.h"
+
+#include "../../../src/include/tscout/marker.h"
```
You could do `tscout/marker.h` instead?

```suggestion
#include "tscout/marker.h"
```

karthik-ramanathan-3006 replied:
Oops. Yes! Thanks for this!

lmwnshn commented on cmudb/extensions/tscout/main.c:
```diff
@@ -0,0 +1,224 @@
+
+static ExplainOneQuery_hook_type chain_ExplainOneQuery = NULL;
+
+void _PG_init(void) {
```
As a general style nitpick, I prefer all comments and messages to be punctuated.

lmwnshn commented on cmudb/extensions/tscout/main.c:
```diff
@@ -0,0 +1,224 @@
+void _PG_init(void) {
+  elog(LOG, "Initializing extension");
+
+  // Chain the Extension Explain Query Wrapper to the head of the chain.
```
Dumb comment: "Extension Explain Query Wrapper" is a mouthful.
If you want to name this extension, feel free to do that, then refer to it by name.
Similarly you could rename `main.c` to be `whatevernameyoupicked.c`.

karthik-ramanathan-3006 replied:
I've rephrases it slightly.
Leaving this open till we rename the extension.
Proposed name: Hutch

lmwnshn replied:
Hutch sounds good to me!

lmwnshn commented on cmudb/extensions/tscout/main.c:
```diff
@@ -0,0 +1,224 @@
+    // This calls into initPlan() which populates the plan tree.
+    // TODO: Create a hook to executor start.
+
+    // Finally, walk through the plan, dumping its out in a separate top-level group.
```
Style nit: I find it helpful to try to avoid using words like `it` in my comments. Instead, prefer the name of the relevant object, if any. This removes ambiguity as to what "it" refers to.

```suggestion
    // Finally, walk through the plan, dumping the output of this extension into a separate top-level group.
```

lmwnshn commented on cmudb/extensions/tscout/main.c:
```diff
@@ -0,0 +1,224 @@
+}
+
+/**
+ * @brief Fetches the size of the field
```
Style nit: please describe a function from the perspective of the caller. Suppose your caller has:

```
// comment
get_field_size(T_BOOL);
```

Try replacing "comment" with "Fetch the size of the field" vs "Fetches the size of the field".

```suggestion
 * @brief Fetch the size of the field.
```

lmwnshn commented on cmudb/extensions/tscout/main.c:
```diff
@@ -0,0 +1,224 @@
+      break;
+  }
+
+  return 0;
```
We don't currently enforce `clang-tidy` on our codebase because a significant part of it is mixed with PostgreSQL.

That said, `clang-tidy main.c -- -I$(../../../build/bin/pg_config --includedir-server)` warns that this can cause a division by zero, which it will, if it gets hit.

I don't think we have standardized on our C error handling strategy, but I would probably explicitly call `abort()` instead of waiting for it to segfault.

karthik-ramanathan-3006 replied:
Thanks. I was wondering what to do in such cases. Will follow this guideline.

lmwnshn commented on cmudb/extensions/tscout/main.c:
```diff
@@ -0,0 +1,224 @@
+ * @param es (ExplainState *) - The current EXPLAIN state.
+ */
+static void explainXs(Plan *node, ExplainState *es) {
+  char *nodeTagExplainer = NULL, nodeName[17];
```
Style nit: I would separate out lines that initialize from lines that just declare.

lmwnshn commented on cmudb/extensions/tscout/main.c:
```diff
@@ -0,0 +1,224 @@
+static void explainXs(Plan *node, ExplainState *es) {
+  char *nodeTagExplainer = NULL, nodeName[17];
+
+  int i, start_index = 0, field_size, next_field_size, padding, num_fields;
```
Ditto above comment for initialization vs declaration.

lmwnshn commented on cmudb/extensions/tscout/main.c:
```diff
@@ -0,0 +1,224 @@
+ * @param es (ExplainState *) - The current EXPLAIN state.
+ */
+static void WalkPlan(Plan *plan, ExplainState *es) {
+  if (plan == NULL) return;
```
1. Will this ever be called on a NULL input?
2. Should this ever be called on a NULL input? Can this be prevented?

If the answer is no to the first question, can this be an `Assert` instead (defined by PostgreSQL)?

lmwnshn commented on cmudb/extensions/tscout/operating_unit_codegen.c:
```diff
@@ -0,0 +1,44 @@
+/**
+ * @brief - An enum that represents all the C data types
+ * encountered while scanning the Postgres code-base.
+ *
```
Please add documentation: where do these types come from? I believe they are from the clang parser, is that correct?

In general, when defining a constant that depends on XYZ, please specify what XYZ is and where it came from.

lmwnshn commented on cmudb/extensions/tscout/operating_unit_codegen.c:
```diff
@@ -0,0 +1,44 @@
+} c_type;
+
+/**
+ * @brief - A struct that represents an X or a feature.
```
Nit: an X _or_ a feature?

```suggestion
 * @brief - A struct that represents a feature for the behavior models (also called an X).
```

lmwnshn commented on cmudb/extensions/tscout/main.c:
```diff
@@ -0,0 +1,224 @@
+static void ExplainOneQueryWrapper(Query *query, int cursorOptions, IntoClause *into, ExplainState *es,
+                                   const char *queryString, ParamListInfo params, QueryEnvironment *queryEnv);
+static void WalkPlan(Plan *plan, ExplainState *es);
+static void explainXs(Plan *node, ExplainState *es);
```
Please be consistent with function naming. As a rough style guide, in the old NoisePage project, we preferred `FunctionNamesToBeLikeThis()`.

```suggestion
static void ExplainFeatures(Plan *node, ExplainState *es);
```

lmwnshn commented on cmudb/extensions/tscout/main.c:
```diff
@@ -0,0 +1,224 @@
+                                   const char *queryString, ParamListInfo params, QueryEnvironment *queryEnv);
+static void WalkPlan(Plan *plan, ExplainState *es);
+static void explainXs(Plan *node, ExplainState *es);
+static size_t get_field_size(c_type type);
```
```suggestion
static size_t GetFieldSize(c_type type);
```

lmwnshn commented on cmudb/extensions/tscout/main.c:
```diff
@@ -0,0 +1,224 @@
+static ExplainOneQuery_hook_type chain_ExplainOneQuery = NULL;
+
+void _PG_init(void) {
+  elog(LOG, "Initializing extension");
```
```suggestion
  elog(LOG, "Initializing extension.");
```

lmwnshn commented on cmudb/extensions/tscout/main.c:
```diff
@@ -0,0 +1,224 @@
+  chain_ExplainOneQuery = ExplainOneQuery_hook;
+  ExplainOneQuery_hook = ExplainOneQueryWrapper;
+
+  // Init logic (like setting flags) go here
```
```suggestion
  // Init logic (like setting flags) go here.
```

lmwnshn commented on cmudb/extensions/tscout/main.c:
```diff
@@ -0,0 +1,224 @@
+void _PG_fini(void) {
+  // Clean up logic goes here
+  ExplainOneQuery_hook = chain_ExplainOneQuery;
+  elog(DEBUG1, "Finishing extension");
```
```suggestion
  elog(DEBUG1, "Finishing extension.");
```

lmwnshn commented on cmudb/extensions/tscout/main.c:
```diff
@@ -0,0 +1,224 @@
+}
+
+void _PG_fini(void) {
+  // Clean up logic goes here
```
```suggestion
  // Clean up logic goes here.
```

lmwnshn commented on cmudb/extensions/tscout/main.c:
```diff
@@ -0,0 +1,224 @@
+  }
+
+  // Postgres does not expose an interface to call into the standard ExplainOneQuery.
+  // Hence, we duplicate the operations performed by the standard ExplainOneQuery ie
```
```suggestion
  // Hence, we duplicate the operations performed by the standard ExplainOneQuery, i.e.,
```

lmwnshn commented on cmudb/extensions/tscout/main.c:
```diff
@@ -0,0 +1,224 @@
+      eflags = EXEC_FLAG_EXPLAIN_ONLY;
+    if (into) eflags |= GetIntoRelEFlags(into);
+
+    // Run the executor
```
```suggestion
    // Run the executor.
```

lmwnshn commented on cmudb/extensions/tscout/main.c:
```diff
@@ -0,0 +1,224 @@
+    FreeQueryDesc(queryDesc);
+  }
+
+  // Finally, after performing the extension specific operations, explain the planner.
```
```suggestion
  // Finally, after performing the extension-specific operations, run the standard explain code path.
```

lmwnshn commented on cmudb/extensions/tscout/main.c:
```diff
@@ -0,0 +1,224 @@
+/**
+ * @brief Fetches the size of the field
+ *
+ * @param type (c_type) - The C field type
```
```suggestion
 * @param type (c_type) - The C field type.
```

lmwnshn commented on cmudb/extensions/tscout/main.c:
```diff
@@ -0,0 +1,224 @@
+ * @brief Fetches the size of the field
+ *
+ * @param type (c_type) - The C field type
+ * @return size_t - Size of the field on the machine
```
```suggestion
 * @return size_t - Size of the field on the machine.
```

lmwnshn commented on cmudb/extensions/tscout/main.c:
```diff
@@ -0,0 +1,224 @@
+ * @param type (c_type) - The C field type
+ * @return size_t - Size of the field on the machine
+ */
+size_t get_field_size(c_type type) {
```
```suggestion
size_t GetFieldSize(c_type type) {
```

lmwnshn commented on cmudb/extensions/tscout/main.c:
```diff
@@ -0,0 +1,224 @@
+}
+
+/**
+ * @brief - Explains the features of the given node.
```
```suggestion
 * @brief - Explain the features of the given node.
```

lmwnshn commented on cmudb/extensions/tscout/main.c:
```diff
@@ -0,0 +1,224 @@
+  ExplainPropertyText("tag", nodeTagExplainer, es);
+
+  for (i = 0; i < num_fields; i++) {
+    field_size = get_field_size(fields[i].type);
```
```suggestion
    field_size = GetFieldSize(fields[i].type);
```

lmwnshn commented on cmudb/extensions/tscout/main.c:
```diff
@@ -0,0 +1,224 @@
+
+  for (i = 0; i < num_fields; i++) {
+    field_size = get_field_size(fields[i].type);
+    next_field_size = i < num_fields - 1 ? get_field_size(fields[i + 1].type) : 8;
```
```suggestion
    next_field_size = i < num_fields - 1 ? GetFieldSize(fields[i + 1].type) : 8;
```

lmwnshn commented on cmudb/extensions/tscout/main.c:
```diff
@@ -0,0 +1,224 @@
+}
+
+/**
+ * @brief - Walks through the plan tree.
```
```suggestion
 * @brief - Walk through the plan tree, dumping features to the current open group.
```

lmwnshn commented on cmudb/extensions/tscout/main.c:
```diff
@@ -0,0 +1,224 @@
+  if (plan == NULL) return;
+
+  // 1. Explain the current node.
+  explainXs(plan, es);
```
```suggestion
  ExplainFeatures(plan, es);
```

lmwnshn commented on cmudb/extensions/tscout/operating_unit_codegen.c:
```diff
@@ -0,0 +1,44 @@
+} field;
+
+/**
+ * @brief - Represents a TScout Operating Unit
```
This is where being named differently from the actual TScout would be handy...

```suggestion
 * @brief - An OperatingUnit as defined by TScout.
```

lmwnshn commented on cmudb/extensions/tscout/tscout_extension--0.0.1.sql:
```diff
@@ -0,0 +1 @@
+-- Nothing to do here
```
```suggestion
-- Nothing to do here.
```

lmwnshn commented on cmudb/extensions/tscout/tscout_extension.control:
```diff
@@ -0,0 +1,4 @@
+# TScout extension
+comment = 'TODO: (Karthik) Enter description here'
```
```suggestion
comment = 'Support for EXPLAIN (FORMAT TSCOUT) to gather query features.'
```

lmwnshn commented on cmudb/extensions/tscout/tscout_extension.control:
```diff
@@ -0,0 +1,4 @@
+# TScout extension
```
```suggestion
# TScout extension.
```

lmwnshn commented on cmudb/extensions/tscout/tscout_feature_gen.py:
```diff
@@ -0,0 +1,160 @@
+
+from clang.cindex import TypeKind
+
+# We're assuming that this script is housed in postgres/cmudb/extensions/tscout
```
```suggestion
# We're assuming that this script is housed in `postgres/cmudb/extensions/tscout`.
```

lmwnshn commented on cmudb/extensions/tscout/tscout_feature_gen.py:
```diff
@@ -0,0 +1,160 @@
+CODEGEN_TEMPLATE_PATH = Path.joinpath(TSCOUT_EXTENSION_PATH, "operating_unit_codegen.c")
+CODEGEN_FILE_PATH = Path.joinpath(TSCOUT_EXTENSION_PATH, "operating_unit_features.h")
+
+TSCOUT_PATH = Path.joinpath(TSCOUT_EXTENSION_PATH, Path("../../tscout/"))
```
Hm. I don't know how I feel about this approach long-term, but I am OK with it for now. Some drawbacks of this approach include losing IDE support.

Perhaps we should make all of the `cmudb` folder a single Python module, but modules can get annoying. Again, I'm OK with this for now -- just documenting my thoughts.

karthik-ramanathan-3006 replied:
I completely agree!
Didn't feel right to add this file to the same directory as Tscout, just to retain IDE support. 
Felt that this script (at this stage) was simple enough to get by, without the IDE.
Long term solution for this would be to restructure our modules, as you suggested.

lmwnshn replied:
Yea, let's just leave this conversation open and we can come back to it in the future

lmwnshn commented on cmudb/extensions/tscout/tscout_feature_gen.py:
```diff
@@ -0,0 +1,160 @@
+TSCOUT_PATH = Path.joinpath(TSCOUT_EXTENSION_PATH, Path("../../tscout/"))
+sys.path.append(TSCOUT_PATH.resolve().__str__())
+
+# And now, we can import TScout
```
```suggestion
# And now, we can import TScout.
```

lmwnshn commented on cmudb/extensions/tscout/tscout_feature_gen.py:
```diff
@@ -0,0 +1,160 @@
+    "statement_timestamp"
+]
+
+def aggregate_features(ou: model.OperatingUnit):
```
We generally use and prefer the numpy docstring format. I went ahead and made these changes.

lmwnshn commented on cmudb/tscout/model.py:
```diff
@@ -677,3 +677,13 @@ def __init__(self):
+            return {}
+
+        return {
+            entry[0]: entry[1] for entry in self._enums.get(enum_name, [])
```
Since you already checked up above, you should never reach the default `[]` case.

```suggestion
            entry[0]: entry[1] for entry in self._enums[enum_name]
```

lmwnshn commented on cmudb/tscout/model.py:
```diff
@@ -677,3 +677,13 @@ def __init__(self):
         self.metrics = OU_METRICS
+        self._enums = nodes.enum_map
+
+    def get_enum_value_map(self, enum_name):
```
Please add a numpy docstring on what this function returns.

lmwnshn commented on cmudb/tscout/clang_parser.py:
```diff
@@ -177,6 +189,11 @@ def __init__(self):
             for node_name in self._bases
         }
 
+        self.enum_map: Mapping[str, List[Tuple]] = {
```
AFAIK that tuple is a str and an int?

```suggestion
        self.enum_map: Mapping[str, List[Tuple[str, int]]] = {
```

lmwnshn commented on cmudb/extensions/tscout/tscout_feature_gen.py:
```diff
@@ -0,0 +1,160 @@
+
+            else:
+                # Print defaults
+                ou_string = fill_in_template(ou_string, -1, "", [])
```
This doesn't typecheck in the third argument.

lmwnshn replied:
Or more accurately, the third argument had the wrong type specified. Fixed.

lmwnshn commented on cmudb/extensions/tscout/tscout_feature_gen.py:
```diff
@@ -0,0 +1,160 @@
+        features_list_string = ""
+        # For each OU, generate the features and fill in the autogeneration template.
+        for (key, value) in OU_TO_FEATURE_LIST_MAP.items():
+            ou_string = match[:] # Copy the matching string by value
```
Python strings are immutable, I don't think you need to do this.
Maybe you had it as a list before?

https://stackoverflow.com/a/44863892

lmwnshn commented on cmudb/extensions/tscout/tscout_feature_gen.py:
```diff
@@ -0,0 +1,160 @@
+    with open(str(CODEGEN_TEMPLATE_PATH), "r") as template:
+        text = template.read()
+
+        matches = re.findall(r"\(ou\)\{.*\}\,", text)                      # Find a seq. that matches "(ou){.*},"
```
There are some unnecessary escapes here (as highlighted by CLion). Going ahead and removing them.

lmwnshn commented on cmudb/extensions/tscout/tscout_feature_gen.py:
```diff
@@ -0,0 +1,160 @@
+# And now, we can import TScout
+from tscout import model
+
+OU_TO_FEATURE_LIST_MAP = {}
```
Nit: when initializing an empty dictionary, please include a type annotation.

lmwnshn replied:
Is this a `Mapping[int, Mapping[str, Any]]`?

lmwnshn replied:
It might be cleaner to define the last `Mapping[str, Any]` as a dataclass, examples in `clang_parser.py`.

lmwnshn commented on cmudb/extensions/tscout/tscout_feature_gen.py:
```diff
@@ -0,0 +1,160 @@
+OU_EXCEPTIONS = {
+    "ExecHashJoinImpl": "T_HashJoin"
+}
+OU_EXCLUDED_FEATURES = [
```
Please add a comment detailing why features are excluded. This comment should help future readers decide whether to add new features to be excluded, or to remove old features from being excluded.

karthik-ramanathan-3006 replied:
Missed adding this. Thank you!

lmwnshn commented on cmudb/extensions/tscout/tscout_feature_gen.py:
```diff
@@ -0,0 +1,160 @@
+    "query_id",
+    "left_child_plan_node_id",
+    "right_child_plan_node_id",
+    "statement_timestamp"
```
```suggestion
    "statement_timestamp",
```

Just future proofing.

lmwnshn committed:
Add numpy docstring and punctuation.

The only content change is adding a punctuation mark to operating_unit_codegen
because that sentence is a find and replace target.

Also fix a mismatched type in fill_in_template.

lmwnshn commented on cmudb/extensions/tscout/tscout_feature_gen.py:
```diff
@@ -0,0 +1,208 @@
+    new_ou_string : str
+        The codegen template with the given OU's details substituted.
+    """
+    # Replace the index of the OU.
```
General comment: we usually have comments on a separate line (you had it to the right).

lmwnshn commented on cmudb/extensions/tscout/tscout_feature_gen.py:
```diff
@@ -0,0 +1,208 @@
+    # If there are features, add the list of features.
+    # Otherwise, replace with a dummy string.
+    if ou_xs:
+        ou_string = ou_string.replace("OU_Xs", "feat_%d" % (ou_index))
```
This is optional: you can use f-strings if you think it makes it more readable

```
f"feat_{ou_index}"
```

In this particular case and for this file in general, I have no strong opinion, only a slight unjustified preference for f-strings. Note that we will still always want to do `%` substitution for logging because that can defer evaluation.

lmwnshn reviewed (changes requested):
Generally LGTM, **very** high code quality and good taste in code. Especially considering the many moving parts across multiple programming languages and the codegen aspect, easily one of the best first-time PRs that I have ever reviewed!

If we were in a rush, I would be perfectly comfortable merging this code as it is.

However, because it is your first PR and because I think you will absorb this, I am reviewing this PR at maximum nitpick level. The majority of my comments and suggestions are architecture- and style-based comments. I recognize that we have not published a formal style guide yet, and we additionally cannot use clang-tidy to enforce it like we did in our NoisePage repo because of all the grungy PostgreSQL code, so you will have to somewhat absorb our current style passively, sorry. To this end, I'm using the PR suggestions instead of just pushing my fixes directly.

---

High-level comments.

Architecture / code:
- Wherever possible, try to engineer functions to fail immediately instead of silently, e.g., abort before it segfaults; assert instead of returning a possibly invalid default value.
- Python strings are immutable, I don't think you need to `copy[:]` them? I could be wrong.
- You don't need to do anything for this: I am not super happy about adding to the PATH for the Python script, but modules are a discussion for the group for the future.

Review process:
- Please include a more descriptive PR comment in the future (feel free to write a placeholder, usually I write this at the time of asking for review).

Style:
- Please separate initialization from declaration, e.g., `int a, b=0, c;` -- put `b=0` on its own line.
- Please use [numpy-style](https://numpydoc.readthedocs.io/en/latest/format.html) docstrings for Python. This is primarily because we interface mostly with numpy / ML code which uses that convention already. This also pushes the type into the documentation instead of the function signature, which IMO makes it easier to read.
- In Python, if you have complicated structures look out for opportunities to simplify them with a `@dataclass`, e.g., `OU_TO_FEATURE_LIST_MAP : Mapping[int, FeatureList]` instead of `Mapping[int, Mapping[str, Any]]`.
- In writing documentation or defining constants, please consider the future reader, e.g., when having a list of exclusions please describe why things are excluded if it is not necessarily obvious; try to avoid using words like "it" in comments because I find that "it" tends to always end up being ambiguous; functions are generally better described from the caller's perspective.
- This is just Wan being Wan: please punctuate comments and do a quick pass for grammar and typos (if you use CLion, a grammar checker is built in and it can catch typos like "the the" which you had). For the most part, you have done this already and maybe just missed a few.

I _hope_ accepting all code suggestions will not break the functionality of your code, but if it does, let me know. Also in general, everything is always open for debate: feel free to push back on basically anything that I suggest / roll back any of the changes that I pushed directly. You can easily see what my changes are by clicking on the "Files changed" -> "Changes from all commits" dropdown tab in GitHub.

---

Again, great work! Even for a non-first PR, this would be impressive.

lmwnshn commented:
One last comment: maybe add a README to the extension folder? Generally, it is preferred that instructions and documentation live inside the codebase (which we'll see often) instead of outside on GitHub (which we only check when we're desperate to debug).

karthik-ramanathan-3006 committed:
Addressed Wan's comments, ran Python linters through cmudb/extensions and cmudb/tscout.

karthik-ramanathan-3006 committed:
Converted the value of OU_TO_FETAURE_LIST_MAP in tscout_feature_gen to a Dataclass. Rant linter

karthik-ramanathan-3006 committed:
Renamed extension to 'Hutch'. Woof!

karthik-ramanathan-3006 committed:
Fixed ExplainFeature function call which was still refering to old name.

karthik-ramanathan-3006 committed:
Revert "Renamed extension to 'Hutch'. Woof!"

This reverts commit 588b1bd5c300fccfe64e39ef26197b6bb2fe95aa.

karthik-ramanathan-3006 committed:
Renamed Postgres extension to 'Hutch'. Woof!

karthik-ramanathan-3006 committed:
Merge branch 'upstream/pg14' into tscout_ext

karthik-ramanathan-3006 committed:
Updated build Makefiles to reflect new name of extension.

lmwnshn commented on cmudb/extensions/hutch/main.c:
```diff
@@ -0,0 +1,227 @@
+}
+
+/**
+ * @brief - Walks through the plan tree, dumping features to the current open group.
```
```suggestion
 * @brief - Walk through the plan tree, dumping features to the current open group.
```

lmwnshn commented on cmudb/extensions/hutch/operating_unit_codegen.c:
```diff
@@ -0,0 +1,45 @@
+/**
+ * @brief - An enum that represents all the C data types
+ * encountered while scanning the Postgres code-base.
+ * The scan operation is performed by the Clang parser.
```
```suggestion
 * The scan operation is performed by the Clang parser.
 * See `cmudb/tscout/clang_parser.py`.
```

lmwnshn commented on cmudb/extensions/hutch/operating_unit_codegen.c:
```diff
@@ -0,0 +1,45 @@
+} c_type;
+
+/**
+ * @brief - a feature for the behavior models (also called an X).
```
```suggestion
 * @brief - A feature for the behavior models (also called an X).
```

lmwnshn commented on cmudb/extensions/hutch/tscout_feature_gen.py:
```diff
@@ -0,0 +1,232 @@
+    ou_name: str
+        The name of the Operating Unit.
+    features: List[Tuple[str, TypeKind]]
+        The list of (feature-name, feature-data-type) pairs which correspond to the features
```
```suggestion
        The list of (feature-name, feature-data-type) pairs which correspond to the features.
```

lmwnshn committed:
Reduce diff by unformatting most of cmudb/tscout.

I hope I didn't revert anything new...

lmwnshn reviewed (approved):
LGTM! Thanks for addressing all nits.

lmwnshn committed:
Small capitalization nitpick.

lmwnshn committed:
Small punctuation nitpick.

lmwnshn committed:
Explicitly name the relevant file.

lmwnshn committed:
Function description style.
